rllib/algorithms/dt/dt_torch_model.py

```diff
@@ -0,0 +1,210 @@
+
+        # Update view requirement
+        # NOTE: See DTTorchPolicy.action_distribution_fn for an explanation of
+        # why the ViewRequirements are like this
+        self.view_requirements = {
+            SampleBatch.OBS: ViewRequirement(
+                space=obs_space, shift=f"-{self.max_seq_len-1}:0"
+            ),
+            SampleBatch.ACTIONS: ViewRequirement(
+                space=action_space, shift=f"-{self.max_seq_len-1}:-1"
+            ),
+            SampleBatch.REWARDS: ViewRequirement(shift=-1),
+            SampleBatch.T: ViewRequirement(shift=f"-{self.max_seq_len-2}:0"),
+            SampleBatch.RETURNS_TO_GO: ViewRequirement(
+                shift=f"-{self.max_seq_len-1}:-1"
+            ),
+        }
+
+    def build_transformer(self):
+        # build the model
+        gpt_config = GPTConfig(
+            block_size=self.block_size,
+            n_layer=self.model_config["num_layers"],
+            n_head=self.model_config["num_heads"],
+            n_embed=self.embed_dim,
+            embed_pdrop=self.model_config["embed_pdrop"],
+            resid_pdrop=self.model_config["resid_pdrop"],
+            attn_pdrop=self.model_config["attn_pdrop"],
+        )
+        gpt = GPT(gpt_config)
+        return gpt
+
+    def build_position_encoder(self):
+        return nn.Embedding(self.max_ep_len, self.embed_dim)
+
+    def build_action_encoder(self):
+        if isinstance(self.action_space, Discrete):
+            return nn.Embedding(self.action_dim, self.embed_dim)
+        elif isinstance(self.action_space, Box):
+            return nn.Linear(self.action_dim, self.embed_dim)
+        else:
+            raise NotImplementedError
+
+    def build_obs_encoder(self):
+        return nn.Linear(self.obs_dim, self.embed_dim)
+
+    def build_return_encoder(self):
+        return nn.Linear(1, self.embed_dim)
+
+    def build_action_head(self):
+        return nn.Linear(self.embed_dim, self.action_dim)
+
+    def build_obs_head(self):
+        if not self.model_config["use_obs_output"]:
+            return None
+        return nn.Linear(self.embed_dim, self.obs_dim)
+
+    def build_return_head(self):
+        if not self.model_config["use_return_output"]:
+            return None
+        return nn.Linear(self.embed_dim, 1)
+
+    @override(ModelV2)
+    def forward(
+        self,
+        input_dict: Dict[str, TensorType],
+        state: List[TensorType],
+        seq_lens: TensorType,
+    ) -> (TensorType, List[TensorType]):
+        # True No-op forward method.
+        # TODO: Support image observation inputs
+        return input_dict["obs"], state
+
+    def get_prediction(
+        self,
+        model_out: TensorType,
+        input_dict: SampleBatch,
+    ) -> Dict[str, TensorType]:
+        """Computes the output of a forward pass of the decision transformer.
+
+        Args:
+            model_out: output observation tensor from the base model, [B x T x obs_dim]
```

Comment:
for shape use [B,T,obs_dim] to avoid confusion with [BxTxobs_dim]

Reply:
done